usersync/ejector.go

```diff
@@ -0,0 +1,134 @@
+package usersync
+
+import (
+	"errors"
+	"math"
+	"time"
+)
+
+type Ejector interface {
+	Choose(uids map[string]UIDEntry) (string, error)
+}
+
+type OldestEjector struct {
+	nonPriorityKeys []string
+	PriorityGroups  [][]string
+	EjectPriority   bool
+}
+
+type PriorityBidderEjector struct {
+	PriorityGroups   [][]string
+	SyncerKey        string
+	OldestEjector    OldestEjector
+	IsSyncerPriority bool
+}
```

Comment:
I should remove the `SyncerKey` since it's no longer used.

Originally I had the `PriorityEjector` call `IsSyncerPriority()` to check if the given syncer key we're trying to sync is a priority as it's part of the requirements to ensure that this is the case. However, I realized it was inefficient to call this function every time the Choose() was called. So instead, I pulled the function up a level, so that it was only called once, and then pass the result as a boolean to the ejector which is what this `IsSyncerPriority` is. But thinking about it now, I feel like if `IsSyncerPriority()` returns false, I should just return an error immediately within `setuid.go`, which would allow me to remove this logic from the ejector entirely.

And the `OldestEjector` is called directly in one place `pbs/usersync.go/OptOut()`. It's called directly since this OptOut method doesn't include any reference to a `SyncerKey`, and thus I can't use the `PriorityEjector` in this case.